### PR TITLE
Added topologySpreadConstraints

### LINKED ISSUE #362: Support passing topologySpreadConstraints to deployment
In addition to `affinity` it would be nice to be able to specify `topologySpreadConstraints`. For example as done in nginx ingress https://github.com/kubernetes/ingress-nginx/pull/6154

### PR DESCRIPTION
**Pull request by SBe**
resolves #362 
Added possibility to specify topologySpreadConstraints (enabled since Kubernetes 1.19).

### CODE REVIEW

File: traefik/templates/_podtemplate.tpl
Comment (SantoDE) [inline traefik/templates/_podtemplate.tpl:23]: Shouldnt that be behind a version check? What will happen on installations prior to 1.19?

**Review by SantoDE — CHANGES_REQUESTED** [review]
Hey 👋 

Thanks for your interest and sorry for the late answer. Days were pretty busy.

One question left, see below :)

File: traefik/templates/_podtemplate.tpl
Comment (SBe) [reply traefik/templates/_podtemplate.tpl]: I would like to add a check `(semverCompare ">=1.19.0" .Capabilities.KubeVersion.Version)` but helm unittest plugin does not support setting .Capabilities and a lot of tests fail. My idea was that someone, who wants to use this feature should now whether or not it is enabled on his/her cluster

Comment (SantoDE) [reply traefik/templates/_podtemplate.tpl:23]: Do you know what happens when someone "accidentally" activates that? Like what is the potential error?

Comment (SBe) [reply traefik/templates/_podtemplate.tpl:23]: It will fail during installation/upgrade

Comment (SantoDE) [reply traefik/templates/_podtemplate.tpl:23]: Then, I think, we should rather hide it behind a version compare and don't have a test for it. >/ As sad as it is, I think better user experience is to favor over better  test coverage on that part.

**Comment by SantoDE** [comment]
Tests are failing @SBe 

Do you manage to fix them? :)

**Comment by SBe** [comment]
Sorry, I was also pretty busy. 
As I mentioned. The problem is the helm-unittest plugin and the fact that it does not support setting Capabilites.
I can delete my test, but the reference to .Capabilities is still in helm chart and other tests fail to render the chart properly (see log below). For now, I can restore the test and make it mergable. If you have any suggestion how to resolve it, I can implement something, but I saw that it would be probably a while until this functionality will be merged into helm-unittest plugin.

`Error: render error in "traefik/templates/deployment.yaml": template: traefik/templates/_podtemplate.tpl:23:92: executing "traefik.podTemplate" at <.Capabilities.KubeVersion.Version>: can't evaluate field Version in type *version.Info`

**Review by jakubhajek — COMMENTED** [review]
Hello Szymon 👋 

The topology spread constraints is a great feature that allows spreading pods across cluster and increase HA and more efficient resource utilization. 

However, we have to make sure that Traefik installed through Helm Chart can be installed on the older version of K8S so backward compatibility is crucial. 

If the change is not testable due to lack of the `Capabilites` feature  [#88](https://github.com/lrills/helm-unittest/issues/88)  in the Helm Unit-test framework I incline to hold on to that PR until we will be sure that we keep backward compatibility.

File: traefik/values.yaml
Comment (SantoDE) [inline traefik/values.yaml:416]:
```suggestion
topologySpreadConstraint
```

File: traefik/values.yaml
Comment (jakubhajek) [inline traefik/values.yaml:418]: The current label is deprecated, I suggest to use the following
```suggestion
#     topologyKey: topology.kubernetes.io/zone
```
https://kubernetes.io/docs/reference/labels-annotations-taints/#topologykubernetesiozone

**Review by jakubhajek — CHANGES_REQUESTED** [review]
👋 
Again, we've discussed that change internally and we are happy to add it. As I mentioned earlier this is a very useful feature so let's add it. Please review my comments and suggestions and bump the Traefik version in the file `traefik/tests/container-config_test.yaml` and we will merge it. Thank you,

File: traefik/Chart.yaml
Comment (SBe) [reply traefik/Chart.yaml]: Version bumped

File: traefik/Chart.yaml
Comment (SBe) [reply traefik/Chart.yaml]: Applied

File: traefik/values.yaml
Comment (jakubhajek) [inline traefik/values.yaml:419]: Seems that something went wrong and that sections is duplicated.

File: traefik/values.yaml
Comment (jakubhajek) [inline traefik/values.yaml:416]: Seems that something went wrong and the below section has been duplicated.  Can you please try to use the following

```yaml
topologySpreadConstraints:
  - maxSkew: 1
    topologyKey: topology.kubernetes.io/zone
    whenUnsatisfiable: DoNotSchedule
    labelSelector:
      matchLabels:
        app.kubernetes.io/instance: traefik
```

File: traefik/tests/pod-config_test.yaml
Comment (jakubhajek) [inline traefik/tests/pod-config_test.yaml:221]: Can you please remove the deprecated key `failure-domain.beta.kubernetes.io/zone` and replace it to `topology.kubernetes.io/zone`?

**Review by jakubhajek — CHANGES_REQUESTED** [review]
Please have a look at the comments in the code.

File: traefik/tests/pod-config_test.yaml
Comment (devops-ecovadis) [reply traefik/tests/pod-config_test.yaml:221]: Sorry, fogot about this one. Changed

File: traefik/values.yaml
Comment (devops-ecovadis) [reply traefik/values.yaml:416]: Weird, fixed

File: traefik/Chart.yaml
Comment (jakubhajek) [inline traefik/Chart.yaml:5]:
```suggestion
version: 9.18.2
```

**Review by jakubhajek — APPROVED** [review]
Thank you for your kind collaboration. The change looks good for me and seems we are ready to merge it.

**Comment by SBe** [comment]
So can we merge it?

**Comment by jakubhajek** [comment]
> So can we merge it?

Yes, we can merge it but we need to add tests because yesterday our colleagues from Rancher merged that [PR](https://github.com/rancher/helm-unittest/pull/12) - allowing us to use `Capabilities` in our tests.

**Comment by SBe** [comment]
I have added the version check and tests.
We still need to wait for the new release of the plugin :(.
Once it will be released, I will bump it in .travis.yml

**Review by jakubhajek — COMMENTED** [review]
Thanks a lot for adding semverComapre to the newly added feature and for updating tests accordingly.

**Comment by jakubhajek** [comment]
hello @SBe  

The feature `topologySpreadConstraints` is already implemented in #466 so we are closing that. Thanks a lot for your contribution, we really appreciate it.

Thanks for your interest in Traefik and don't hesitate to contribute to #466 or any other PR.

**Pull request closed without merging**
